src/main/java/me/chrr/scribble/mixin/BookEditScreenMixin.java

```diff
@@ -646,4 +804,37 @@ private BookEditScreen.PageContent createPageContent() {
 
         return new RichPageContent(text, cursorPosition, atEnd, lineStartsArray, lines.toArray(new BookEditScreen.Line[0]), selectionRectangles.toArray(new Rect2i[0]));
     }
+
+
+    @Override
+    public BookEditScreenMemento scribble$createMemento() {
+        RichSelectionManager selectionManager = this.getRichSelectionManager();
+        return new BookEditScreenMemento(
+                currentPage,
+                selectionManager.selectionStart,
+                selectionManager.selectionEnd,
+                getCurrentPageText(),
+                activeColor,
+                Set.copyOf(activeModifiers)
+        );
+    }
+
+    @Override
+    public void scribble$restore(BookEditScreenMemento memento) {
```

Comment:
`this.dirty = true`, see above.

Reply:
Restore calls `setPageText`, that already has `this.dirty = true`. So it's not a fix.